File: packages/go/storage.go
package openrouterauto

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
)

// Storage is the minimal key/value interface used by the client.
type Storage interface {
	Get(key string) (any, bool)
	Set(key string, value any)
	Remove(key string)
	Clear()
}

// ── MemoryStorage ─────────────────────────────────────────────────────────

// MemoryStorage is the default in-process, non-persistent store.
type MemoryStorage struct {
	mu    sync.RWMutex
	store map[string]any
}

// NewMemoryStorage returns an initialised MemoryStorage.
func NewMemoryStorage() *MemoryStorage {
	return &MemoryStorage{store: make(map[string]any)}
}

func (m *MemoryStorage) Get(key string) (any, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	v, ok := m.store[key]
	return v, ok
}

func (m *MemoryStorage) Set(key string, value any) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.store[key] = value
}

func (m *MemoryStorage) Remove(key string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.store, key)
}

func (m *MemoryStorage) Clear() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.store = make(map[string]any)
}

// ── FileStorage ───────────────────────────────────────────────────────────

// FileStorage persists state to a JSON file with 0600 permissions.
// The resolved path must fall under CWD or the user's home directory
// to prevent path-traversal attacks.
type FileStorage struct {
	path string
	mu   sync.Mutex
	data map[string]any
}

// NewFileStorage creates a FileStorage at configPath.
// Returns an error when the path resolves outside CWD / HOME.
func NewFileStorage(configPath string) (*FileStorage, error) {
	if configPath == "" {
		configPath = ".openrouter-auto.json"
	}
	abs, err := filepath.Abs(configPath)
	if err != nil {
		return nil, fmt.Errorf("invalid config path: %w", err)
	}
	cwd, _ := os.Getwd()
	home, _ := os.UserHomeDir()
	if !isUnder(abs, cwd) && !isUnder(abs, home) {
		return nil, fmt.Errorf(
			"configPath %q resolves outside CWD and HOME — not allowed to prevent path traversal",
			configPath,
		)
	}
	fs := &FileStorage{path: abs, data: make(map[string]any)}
	_ = fs.load()
	return fs, nil
}

// isUnder reports whether path is equal to or nested inside base.
func isUnder(path, base string) bool {
	rel, err := filepath.Rel(base, path)
	if err != nil {
		return false
	}
	// If rel starts with ".." the path is outside base.
	return len(rel) >= 2 && rel[:2] != ".." || rel == "."
}

func (f *FileStorage) Get(key string) (any, bool) {
	f.mu.Lock()
	defer f.mu.Unlock()
	v, ok := f.data[key]
	return v, ok
}

func (f *FileStorage) Set(key string, value any) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.data[key] = value
	_ = f.save()
}

func (f *FileStorage) Remove(key string) {
	f.mu.Lock()
	defer f.mu.Unlock()
	delete(f.data, key)
	_ = f.save()
}

func (f *FileStorage) Clear() {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.data = make(map[string]any)
	_ = f.save()
}

func (f *FileStorage) load() error {
	b, err := os.ReadFile(f.path)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return err
	}
	return json.Unmarshal(b, &f.data)
}

func (f *FileStorage) save() error {
	b, err := json.MarshalIndent(f.data, "", "  ")
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(f.path), 0700); err != nil {
		return err
	}
	return os.WriteFile(f.path, b, 0600)
}

File: packages/rust/src/storage.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

/// Trait for key-value storage used by the client.
pub trait Storage: Send + Sync {
    fn get(&self, key: &str) -> Option<serde_json::Value>;
    fn set(&self, key: &str, value: serde_json::Value);
    fn remove(&self, key: &str);
    fn clear(&self);
}

// ── MemoryStorage ───────────────────────────────────────────────────────────

pub struct MemoryStorage {
    store: Mutex<HashMap<String, serde_json::Value>>,
}

impl MemoryStorage {
    pub fn new() -> Self {
        Self {
            store: Mutex::new(HashMap::new()),
        }
    }
}

impl Default for MemoryStorage {
    fn default() -> Self {
        Self::new()
    }
}

impl Storage for MemoryStorage {
    fn get(&self, key: &str) -> Option<serde_json::Value> {
        self.store.lock().unwrap().get(key).cloned()
    }
    fn set(&self, key: &str, value: serde_json::Value) {
        self.store.lock().unwrap().insert(key.to_string(), value);
    }
    fn remove(&self, key: &str) {
        self.store.lock().unwrap().remove(key);
    }
    fn clear(&self) {
        self.store.lock().unwrap().clear();
    }
}

// ── FileStorage ─────────────────────────────────────────────────────────────

/// Persists state to a JSON file with restricted path traversal.
pub struct FileStorage {
    path: PathBuf,
    data: Mutex<HashMap<String, serde_json::Value>>,
}

impl FileStorage {
    /// Create a FileStorage at the given path.
    /// The resolved path must be under `CWD` or `HOME`.
    pub fn new(config_path: Option<&str>) -> Result<Self, String> {
        let raw = config_path.unwrap_or(".openrouter-auto.json");
        let abs = std::fs::canonicalize(raw)
            .or_else(|_| {
                // File may not exist yet; canonicalize the parent.
                let p = Path::new(raw);
                let parent = p.parent().unwrap_or(Path::new("."));
                let parent_abs = std::fs::canonicalize(parent)
                    .map_err(|e| format!("invalid config path: {}", e))?;
                Ok::<PathBuf, String>(parent_abs.join(p.file_name().unwrap_or_default()))
            })
            .map_err(|e| format!("invalid config path: {}", e))?;

        let cwd = std::env::current_dir().unwrap_or_default();
        let home = dirs_path();
        if !is_under(&abs, &cwd) && !is_under(&abs, &home) {
            return Err(format!(
                "configPath {:?} resolves outside CWD and HOME — not allowed to prevent path traversal",
                raw
            ));
        }

        let mut data = HashMap::new();
        if abs.exists() {
            if let Ok(bytes) = std::fs::read(&abs) {
                if let Ok(parsed) = serde_json::from_slice(&bytes) {
                    data = parsed;
                }
            }
        }

        Ok(Self {
            path: abs,
            data: Mutex::new(data),
        })
    }

    fn save(&self) {
        let data = self.data.lock().unwrap();
        if let Ok(bytes) = serde_json::to_vec_pretty(&*data) {
            if let Some(parent) = self.path.parent() {
                let _ = std::fs::create_dir_all(parent);
            }
            let _ = std::fs::write(&self.path, bytes);
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                let _ = std::fs::set_permissions(
                    &self.path,
                    std::fs::Permissions::from_mode(0o600),
                );
            }
        }
    }
}

impl Storage for FileStorage {
    fn get(&self, key: &str) -> Option<serde_json::Value> {
        self.data.lock().unwrap().get(key).cloned()
    }
    fn set(&self, key: &str, value: serde_json::Value) {
        self.data
            .lock()
            .unwrap()
            .insert(key.to_string(), value);
        self.save();
    }
    fn remove(&self, key: &str) {
        self.data.lock().unwrap().remove(key);
        self.save();
    }
    fn clear(&self) {
        self.data.lock().unwrap().clear();
        self.save();
    }
}

fn is_under(path: &Path, base: &Path) -> bool {
    path.starts_with(base)
}

fn dirs_path() -> PathBuf {
    std::env::var("HOME")
        .map(PathBuf::from)
        .unwrap_or_default()
}
